testdata/generate.py

```diff
@@ -0,0 +1,526 @@
+                close_date="{}-{:02d}-01".format(year, month),
+                dont_create_payments=False,
+            )
+
+        self.writer.writerow(attrs)
+        return attrs
+
+
+class HouseholdGenerator(Generator):
+    table = "accounts"
+    columns = ("id", "name", "record_type")
+
+    def __init__(self, conn, logger, include_skew=False):
+        super(HouseholdGenerator, self).__init__(conn, logger)
+        self.include_skew = include_skew
+
+    def init_subgenerators(self):
+        opportunity_generator = OpportunityGenerator(
+            self.conn,
+            self.logger,
+            {0: 2, 1: 3, 2: 2, 3: 2, 4: 1, 5: 1, 6: 1, 7: 1, 8: 1, 9: 1, 10: 1},
+        )
+        self.generators = {
+            "contacts": ContactGenerator(
+                self.conn, self.logger, {1: 10, 2: 10, 3: 1, 4: 1, 5: 1}
+            ),
+            "addresses": AddressGenerator(self.conn, self.logger, {1: 50, 2: 1}),
+            "recurring_donations": RecurringDonationGenerator(
+                self.conn, self.logger, {1: 1, 0: 4}, opportunity_generator
+            ),
+            "opportunities": opportunity_generator,
+        }
+
+    def make_one(self, n):
+        id = self.make_id()
+        attrs = {"id": id, "record_type": "HH_Account"}
+
+        # contacts
+        contact_gen = self.generators["contacts"]
+        num_contacts = contact_gen.choose_count()
+        contacts = []
+        for x in range(num_contacts):
+            contacts.append(contact_gen.make_one(account_id=id))
+        # name household based on contacts
+        attrs["name"] = self.name_household(contacts)
+
+        # addresses (some seasonal)
+        address_gen = self.generators["addresses"]
+        num_addresses = address_gen.choose_count()
+        if num_addresses == 1:
+            address_gen.make_one(household_id=id)
+        else:
+            address_gen.make_one(
+                household_id=id, start_month=6, start_day=1, end_month=8, end_day=31
+            )
+            address_gen.make_one(
+                household_id=id, start_month=9, start_day=1, end_month=5, end_day=31
+            )
+
+        # recurring donations
+        rd_gen = self.generators["recurring_donations"]
+        num_rds = rd_gen.choose_count()
+        if num_rds:
+            rd_gen.make_one(attrs, contacts)
+
+        # one-off opportunities
+        opp_gen = self.generators["opportunities"]
+        if self.include_skew and n == 1:
+            num_opps = 50000
+            attrs["name"] = "50k Opp Household"
+        elif self.include_skew and n < 10:
+            num_opps = 5000
+            attrs["name"] = "5k Opp Household"
+        else:
+            num_opps = opp_gen.choose_count()
+        for x in range(num_opps):
+            opp_gen.make_one(contacts, account_id=id)
+
+        self.writer.writerow(attrs)
+        return attrs
+
+    def make_many(self, count, **kw):
+        i = 0
+        while i < count:
```

Comment:
Why the `while` loop?

Reply:
Huh yeah, I can't remember and will change it.